### PR TITLE
DEX 101 & 426 email sending framework and get API tested

### PR DESCRIPTION
**Pull request by hwindsor**
## Pull Request Overview

- Created framework for email generation, minimal testing performed
- Further testing of Notifications, get now returns the appropriate Notifications for the user.
- Get NotificationById tested which contains the collaboration id for a collaboration request
- The base Notification schema also contains the sender name and email
- Notification config expanded to include a load more details for email and others,
- Notification patching now implemented and tested

---

### CODE REVIEW

**Comment by naknomum** [comment]
my 2 cents:
- i think since Notifications are generic; they should just be left until user explicitly deletes them (if they end up pointing to a defunct collaboration or something, the api/ui/ux should handle that case)
- if "email generation" is meant to be generic also (not Notification-specific), then i think we _do_ need this for mvp for password reset
- yup makes sense to me you pass as many objects (or their guids) to the notification (and thus message to the user) as possible.

File: app/modules/emails/utils.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+                text = re.sub(f'@{key}@', replacements[key], text)
+        return text
+
+    @classmethod
+    def _send_email_with_mailchimp(cls, email_contents):
+        # TODO In future this will link in to the library to do this
+        pass
+
+    @classmethod
+    def send_email(cls, email_contents):
```
Comment (naknomum) [inline app/modules/emails/utils.py:29]: i wonder if this needs added: `recipient`, `subject`, maybe `sender` ?

`subject` is a weird one.  should that be its own template?  i think we hard-coded in old world.  :cry:

File: app/modules/notifications/models.py [hunk 2]
```diff
@@ -75,29 +99,53 @@ def __repr__(self):
...
         return channels
 
+    def send_if_required(self):
+        from app.modules.emails.utils import EmailUtils
+
+        channels = self.channels_to_send(False)
+        if channels[NotificationChannel.email]:
+            outgoing_message = EmailUtils.build_email(
+                self.message_type, self.message_values
+            )
```
Comment (naknomum) [inline app/modules/notifications/models.py:130]: see my comment above about adding `subject` value to `send_email()` ... i wonder if this should be its own (teeny-tiny plaintext) template that is built out like `outgoing_message`?   so we can have things like "_You have a collaboration request from {user} on {sitename}"_ or whatever.

**Review by naknomum — COMMENTED** [review]
just a couple comments around `send_email()`

File: app/modules/emails/utils.py (hunk 1 above)
Comment (hwindsor) [reply app/modules/emails/utils.py:29]: Very good points. 
I think NOTIFICATION_FIELDS needs to be replaced with NOTIFICATION_CONFIG indexed on NotificationType containing 
mandatory_fields
email template filename
email subject.

Should sender_email be a mandatory field for all Notifications? I think it should now.

File: app/modules/notifications/models.py (hunk 2 above)
Comment (hwindsor) [reply app/modules/notifications/models.py:130]: Yes I like the idea, should be easy to do.

File: app/modules/emails/utils.py (hunk 1 above)
Comment (naknomum) [reply app/modules/emails/utils.py:29]: i agree mandatory, even if that means we stuff in some kinda `do-not-reply@wildme.org` or something.  most outgoing mail services wants _a_ sender.

**Comment by naknomum** [comment]
> just a couple comments around `send_email()`

`send_email()` stuff looks good with subject, sender, etc and jinja templates, great!

**Comment by naknomum** [comment]
oops i hit the **close with comment** button incorrectly!

**Review by naknomum — APPROVED** [review]
lgtm

**Pull request merged**